application/models/services/QuestionGroupService.php

```diff
@@ -0,0 +1,434 @@
+
+use LimeSurvey\Models\Services\Proxy\ProxyExpressionManager;
+use LimeSurvey\Models\Services\Proxy\ProxyQuestionGroup;
+use LSYii_Application;
+use Survey;
+use Permission;
+use Question;
+use QuestionGroup;
+use QuestionGroupL10n;
+use Exception;
+use LimeSurvey\Models\Services\Exception\{
+    PersistErrorException,
+    NotFoundException,
+    PermissionDeniedException
+};
+
+/**
+ * @TODO There is a separate service GroupHelper.php whose function(s)
+ * @TODO should be moved here in the future
+ * @SuppressWarnings(PHPMD.ExcessiveClassComplexity)
+ */
+class QuestionGroupService
+{
+    private Permission $modelPermission;
+    private Survey $modelSurvey;
+    private Question $modelQuestion;
+    private QuestionGroup $modelQuestionGroup;
+    private QuestionGroupL10n $modelQuestionGroupL10n;
+    private ProxyExpressionManager $proxyExpressionManager;
+    private ProxyQuestionGroup $proxyQuestionGroup;
+    private LSYii_Application $yiiApp;
+
+    public function __construct(
+        Permission $modelPermission,
+        Survey $modelSurvey,
+        Question $modelQuestion,
+        QuestionGroup $modelQuestionGroup,
+        QuestionGroupL10n $modelQuestionGroupL10n,
+        ProxyExpressionManager $proxyExpressionManager,
+        ProxyQuestionGroup $proxyQuestionGroup,
+        LSYii_Application $yiiApp
+    ) {
+        $this->modelPermission = $modelPermission;
+        $this->modelSurvey = $modelSurvey;
+        $this->modelQuestion = $modelQuestion;
+        $this->modelQuestionGroup = $modelQuestionGroup;
+        $this->modelQuestionGroupL10n = $modelQuestionGroupL10n;
+        $this->proxyExpressionManager = $proxyExpressionManager;
+        $this->proxyQuestionGroup = $proxyQuestionGroup;
+        $this->yiiApp = $yiiApp;
+    }
+
+    /**
+     * Updates a question group and all the languages.
+     *
+     * @param int $surveyId the survey id
+     * @param int $questionGroupId the question group id
+     * @param array $input has the data for a question group, including an array for languages
+     *                     ['questionGroup']
+     *                          [gid]
+     *                          [sid]
+     *                          [group_order]
+     *                          [randomization_group]
+     *                          [grelevance]
+     *                     ['questionGroupI10N']
+     *                         [en]
+     *                            [group_name]
+     *                            [description]
+     *                         [...]    //more languages
+     * @return QuestionGroup
+     * @throws NotFoundException
+     * @throws PermissionDeniedException
+     * @throws PersistErrorException
+     */
+    public function updateGroup(int $surveyId, int $questionGroupId, array $input)
+    {
+        $survey = $this->getSurvey($surveyId);
+
+        if (!$this->modelPermission->hasSurveyPermission($survey->sid, 'surveycontent', 'update')) {
```

Comment:
> The PSR-12 standard doesn't impose a hard limit on line length, but it recommends a soft limit of 120 characters, and to keep lines under 80 characters where possible.

If you look at my service classes. I break code down to keep lines less than 80 characters. This does increase line length but I think it helps read ability. It doesn't make sense to make these changes when making a small change in a file but when creating a bran new class we should follow this.

I wont go through all of the code and make the same comment but this involves updating a lot of lines.